Fixes #32633 - handle scl and nonscl package names dynamically

upadhyeammit opened the pull request:
To restrict the number of executions, different portions of the code employ package installed conditions.
This update eliminates the requirement for explicit package names for plugins by introducing helper methods to generate package names depending on OS versions EL7(scl or non scl), EL8, and Debian.

upadhyeammit commented:
@evgeni request to review, thank you!

evgeni commented:
Looking at the code, this looks like it would work.

But again, design question: Right now you define a helper `scl_or_nonscl_package` (and a `find_…`), which is very EL7 focused and also sort-of a lie. We do use non-SCL packages on EL7 too. Or other SCLs (Redis, PostgreSQL, MongoDB). Even tho no of these are touched in this PR (or even part of the code).

My proposal would be instead have helpers of the kind `plugin_package(plugin_name)`  and `hammer_plugin_package(plugin_name)`, which will yield the correct package for the current OS.

Like `plugin_package('discovery') would yield `rubygem-foreman_discovery` on EL8, `tfm-rubygem-foreman_discovery` on EL7 and `ruby-foreman-discovery` on Debian. See the acceptance tests of `puppet-foreman` for an example: https://github.com/theforeman/puppet-foreman/blob/b193a1465404fb867a8330a783de09d7ce7e793a/spec/classes/cli_plugins_spec.rb#L12-L18

This would make it explicit *what* kind of package we're requesting and not spill other package names with prefixes that are probably wrong for them.

evgeni commented on definitions/features/foreman_tasks.rb:
```diff
@@ -23,8 +23,9 @@ class Features::ForemanTasks < ForemanMaintain::Feature
+      package_name = scl_or_nonscl_package('rubygem-foreman-tasks')
       check_min_version('ruby193-rubygem-foreman-tasks', '0.6') ||
-        check_min_version('tfm-rubygem-foreman-tasks', '0.7')
+        check_min_version(package_name, '0.7')
```
Unrelated, but should this be "check_plugin_version"?

ekohl commented on lib/foreman_maintain/concerns/system_helpers.rb:
```diff
@@ -217,6 +208,37 @@ def el8?
+            tasks: 'rubygem-foreman-tasks' }
+        if el7?
+          "tfm-#{package_names.fetch(name)}"
+        elsif el8? || debian?
+          package_names.fetch(name)
```
The Debian bit is not correct. Debian uses `ruby-` and dashes since underscores are illegal.

So in the installer we have (check `$plugin_prefix`):
https://github.com/theforeman/puppet-foreman/blob/1734ec1774fb09fd3c4f82646cd54deb3651d14d/manifests/params.pp#L88-L104

Then we transform it here:
https://github.com/theforeman/puppet-foreman/blob/1734ec1774fb09fd3c4f82646cd54deb3651d14d/manifests/plugin.pp#L25
https://github.com/theforeman/puppet-foreman/blob/1734ec1774fb09fd3c4f82646cd54deb3651d14d/manifests/plugin.pp#L33-L40

upadhyeammit commented:
Hello @ekohl and @evgeni,

Added changes as per the references. With this I also removed version checks for packages because upon checking those versions those were mostly of 6.1 or 6.2 Satellite.

Request to review, thank you!

evgeni commented on definitions/features/foreman_cockpit.rb:
```diff
@@ -3,7 +3,7 @@ class Features::ForemanCockpit < ForemanMaintain::Feature
 
     confine do
-      server? && find_package('tfm-rubygem-foreman_remote_execution-cockpit')
+      server? && plugin_package_name('remote_execution-cockpit', 'foreman')
```
same, `plugin_package_name` doesn't verify it's installed

evgeni commented on definitions/features/foreman_openscap.rb:
```diff
@@ -3,7 +3,7 @@ class Features::ForemanOpenscap < ForemanMaintain::Feature
 
     confine do
-      check_min_version('tfm-rubygem-foreman_openscap', '0.5.3')
+      plugin_package_name('openscap', 'foreman')
```
same, `plugin_package_name` doesn't verify it's installed

evgeni commented on definitions/features/foreman_tasks.rb:
```diff
@@ -23,8 +23,7 @@ class Features::ForemanTasks < ForemanMaintain::Feature
     confine do
-      check_min_version('ruby193-rubygem-foreman-tasks', '0.6') ||
-        check_min_version('tfm-rubygem-foreman-tasks', '0.7')
+      plugin_package_name('tasks', 'foreman')
```
same, `plugin_package_name` doesn't verify it's installed

evgeni commented on definitions/features/hammer.rb:
```diff
@@ -7,7 +7,7 @@ class Features::Hammer < ForemanMaintain::Feature
     confine do
       # FIXME: How does this run on proxy?
-      find_package('rubygem-hammer_cli') || find_package('tfm-rubygem-hammer_cli')
+      hammer_package
```
same, `hammer_package` doesn't verify it's installed

evgeni commented on definitions/procedures/foreman_docker/remove_foreman_docker.rb:
```diff
@@ -6,11 +6,12 @@ class RemoveForemanDocker < ForemanMaintain::Procedure
 
     def docker_package
-      'tfm-rubygem-foreman_docker'
+      plugin_package_name('docker', 'foreman')
```
same, `plugin_package_name` doesn't verify it's installed

evgeni commented on lib/foreman_maintain/concerns/system_helpers.rb:
```diff
@@ -217,6 +208,60 @@ def el8?
+      end
+
+      def smart_proxy_plugin_prefix
+        if el7?
```
the proxy was SCL'ed in Foreman 2.0, before that, the packages were also `rubygem-smart_proxy…`

evgeni replied:
Of course, if we don't need to support pre-2.0 versions, that's nice ;)

evgeni commented on lib/foreman_maintain/concerns/system_helpers.rb:
```diff
@@ -217,6 +208,60 @@ def el8?
         return os_facts['os']['release']['major'] if el?
       end
+
+      def foreman_plugin_prefix
```
I wonder if the following is more "plugable"

```ruby
def ruby_prefix(scl=true)
  if el7? && scl
    'tfm-rubygem-'
  elsif el7? || el8?
    'rubygem-'
  elsif debian?
    'ruby-'
  end
end

def foreman_plugin_name(plugin)
  if debian?
    plugin.tr!('_', '-')
  end
  ruby_prefix + plugin
end

def proxy_plugin_name(plugin)
  if debian?
    plugin.tr!('_', '-')
    proxy_plugin_prefix = 'smart-proxy-'
  else
    proxy_plugin_prefix = 'smart_proxy_'
  end
  scl = foreman_version >= '2.0'
  ruby_prefix(scl) + proxy_plugin_prefix + plugin
end
```

I intentionally left out the 'foreman' out of `foreman_plugin_name`, as we have at least one plugin that is called `foreman-tasks` (and not `foreman_tasks`)

Not insisting on the above, just giving ideas.

evgeni commented on lib/foreman_maintain/concerns/system_helpers.rb:
```diff
@@ -217,6 +208,60 @@ def el8?
+
+      def foreman_plugin_prefix
+        if el7?
+          'tfm-rubygem-foreman_'
```
As mentioned in the proposal above, we have plugins that use `foreman-tasks` notation. And of course there is also katello, which doesn't use the foreman prefix at all.

ekohl reviewed (commented):
Just a note that for RPMs there are provides:
https://github.com/theforeman/foreman-packaging/blob/2a9cbbd5baba8d4db837d1e608b32dcff6fc9bf9/packages/plugins/rubygem-foreman-tasks/rubygem-foreman-tasks.spec#L42

We have the same thing for Proxy plugins:

https://github.com/theforeman/foreman-packaging/blob/2a9cbbd5baba8d4db837d1e608b32dcff6fc9bf9/packages/plugins/rubygem-smart_proxy_ansible/rubygem-smart_proxy_ansible.spec#L46

These are the same for both EL7 and EL8. If you can check for those, it will be more consistent and probably less code. These do not exist for Debian packages.

evgeni commented:
> Just a note that for RPMs there are provides:
> https://github.com/theforeman/foreman-packaging/blob/2a9cbbd5baba8d4db837d1e608b32dcff6fc9bf9/packages/plugins/rubygem-foreman-tasks/rubygem-foreman-tasks.spec#L42
> 
> We have the same thing for Proxy plugins:
> 
> https://github.com/theforeman/foreman-packaging/blob/2a9cbbd5baba8d4db837d1e608b32dcff6fc9bf9/packages/plugins/rubygem-smart_proxy_ansible/rubygem-smart_proxy_ansible.spec#L46
> 
> These are the same for both EL7 and EL8. If you can check for those, it will be more consistent and probably less code. These do not exist for Debian packages.

Is there a way to ask `rpm` if a `provides` is installed? Currently, the code uses `rpm -q` and that doesn't work for provides.

ekohl commented:
https://unix.stackexchange.com/questions/242092/how-to-query-which-rpm-package-provides-a-specific-version suggests `rpm -q --whatprovides` works. For example:
```console
# rpm -q --whatprovides foreman-plugin-puppetdb
tfm-rubygem-puppetdb_foreman-5.0.0-3.fm2_5.el7.noarch
# echo $?
0
# rpm -q --whatprovides foreman-plugin-doesnotexist
no package provides foreman-plugin-doesnotexist
# echo $?
1
```

upadhyeammit commented:
> https://unix.stackexchange.com/questions/242092/how-to-query-which-rpm-package-provides-a-specific-version suggests `rpm -q --whatprovides` works. For example:
> 
> ```
> # rpm -q --whatprovides foreman-plugin-puppetdb
> tfm-rubygem-puppetdb_foreman-5.0.0-3.fm2_5.el7.noarch
> # echo $?
> 0
> # rpm -q --whatprovides foreman-plugin-doesnotexist
> no package provides foreman-plugin-doesnotexist
> # echo $?
> 1
> ```

Hello @ekohl 

After taking path of using `--whatprovides` I can see there are at least these combination of `Provides` mentioned in plugins, accordingly I feel approach suggested by the @evgeni is simpler?

~~~
foreman-plugin-dhcp-browser
foreman-%{plugin_name}
foreman-%{plugin_name} = %{version}
foreman-proxy-plugin-%{plugin_name} = %{version}
Provides the ability to boot a host into a rescue system
rubygem(%{gem_name}) = %{version}
%{?scl_prefix}rubygem(%{gem_name}) = %{version}
%{?scl_prefix}rubygem(%{gem_name}) = %{version}-%{release}
~~~

ekohl commented:
For now I think @evgeni's solution is the one I'd go with. I've opened https://projects.theforeman.org/issues/33209 for a better long term solution.

evgeni commented on lib/foreman_maintain/concerns/system_helpers.rb:
```diff
@@ -217,6 +208,53 @@ def el8?
+      end
+
+      def hammer_package
+        if el7?
+          'tfm-rubygem-hammer_cli'
+        elsif el8?
+          'rubygem-hammer_cli'
+        elsif debian?
+          'rubygem-hammer-cli'
+        end
```
```suggestion
        if debian?
          hammer_prefix = 'hammer-cli'
        else
          hammer_prefix = 'hammer_cli'
        end
        ruby_prefix + hammer_prefix
```

(the original code used `rubygem` for Debian, which is wrong, and then I was like "wait, we have a helper for that!")

evgeni reviewed (changes requested):
last nitpick!

evgeni commented:
I'm happy!

@ekohl, you?

ekohl commented on definitions/features/foreman_openscap.rb:
```diff
@@ -3,7 +3,7 @@ class Features::ForemanOpenscap < ForemanMaintain::Feature
     label :foreman_openscap
 
     confine do
-      check_min_version('tfm-rubygem-foreman_openscap', '0.5.3')
```
Looks like 0.5.3 was released in January 2016. I'm guessing that's why it's no longer relevant to check for the version.

upadhyeammit replied:
Yes, that's what I thought

ekohl commented on definitions/features/hammer.rb:
```diff
@@ -7,7 +7,7 @@ class Features::Hammer < ForemanMaintain::Feature
     label :hammer
     confine do
       # FIXME: How does this run on proxy?
```
We can answer this: it can run on a proxy, but doesn't by default. Probably safe to remove.

ekohl commented on definitions/procedures/foreman_docker/remove_foreman_docker.rb:
```diff
@@ -6,11 +6,12 @@ class RemoveForemanDocker < ForemanMaintain::Procedure
 
     def docker_package
-      'tfm-rubygem-foreman_docker'
+      find_package(foreman_plugin_name('foreman_docker'))
```
Doesn't this return `nil` if it's not found? There's also already a `return unless` on line 13 so I'd think this is sufficient:
```suggestion
      foreman_plugin_name('foreman_docker')
```

evgeni replied:
Good catch

ekohl commented on definitions/procedures/foreman_docker/remove_foreman_docker.rb:
```diff
@@ -6,11 +6,12 @@ class RemoveForemanDocker < ForemanMaintain::Procedure
     end
 
     def run
       return unless execute?("rpm -q #{docker_package}")
```
Oh, and this should probably use the package manager's `installed?` method instead.

upadhyeammit replied:
removed this and moved condition to confine block

upadhyeammit commented:
@ekohl request to check, thank you!

ekohl commented on definitions/procedures/foreman_docker/remove_foreman_docker.rb:
```diff
@@ -3,16 +3,14 @@ class RemoveForemanDocker < ForemanMaintain::Procedure
-
-    def docker_package
-      'tfm-rubygem-foreman_docker'
+      confine do
+        find_package(foreman_plugin_name('foreman_docker'))
+      end
```
Oh, this looks even nicer to combine.
